src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+        /// Does not handle params parameters.
+        /// </summary>
+        private static async Task<(bool isParameterized, bool hasOptionalParameter)> ShouldExpressionDisplayCodeActionAsync(
+            Document document, TExpressionSyntax expression, IMethodSymbol methodSymbol,
+            CancellationToken cancellationToken)
+        {
+            var variablesInExpression = expression.DescendantNodes().OfType<TIdentifierNameSyntax>();
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var hasOptionalParameter = false;
+
+            foreach (var parameter in methodSymbol.Parameters)
+            {
+                if (parameter.HasExplicitDefaultValue)
+                {
+                    hasOptionalParameter = true;
+                }
+            }
+
+            foreach (var variable in variablesInExpression)
+            {
+                var parameterSymbol = semanticModel.GetSymbolInfo(variable, cancellationToken).Symbol;
+                if (parameterSymbol is not IParameterSymbol parameter)
+                {
+                    return (false, hasOptionalParameter);
+                }
+                if (parameter.IsParams)
+                {
+                    return (false, hasOptionalParameter);
+                }
+            }
+
+            return (methodSymbol != null && methodSymbol.GetParameters().Any(), hasOptionalParameter);
+        }
+
+        /// <summary>
+        /// Creates new code actions for each introduce parameter possibility.
+        /// Does not create actions for overloads/trampoline if there are optional parameters or if the methodSymbol
+        /// is a constructor.
+        /// </summary>
+        private async Task<ImmutableArray<CodeAction>> AddActionsAsync(Document document,
+            TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+            CancellationToken cancellationToken)
+        {
+            var actionsBuilder = new ArrayBuilder<CodeAction>();
+            var (isParameterized, hasOptionalParameter) = await ShouldExpressionDisplayCodeActionAsync(document, expression, methodSymbol, cancellationToken).ConfigureAwait(false);
+            if (isParameterized)
+            {
+                actionsBuilder.Add(CreateNewCodeAction(FeaturesResources.Introduce_parameter_for_0,
+                    document, expression, methodSymbol, containingMethod, allOccurrences: false, trampoline: false, overload: false, cancellationToken));
+                actionsBuilder.Add(CreateNewCodeAction(FeaturesResources.Introduce_parameter_for_all_occurrences_of_0,
+                    document, expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: false, overload: false, cancellationToken));
+
+                if (!hasOptionalParameter && methodSymbol.MethodKind is not MethodKind.Constructor)
```

Comment:
why are we not updating constructors?
what about indexers?

Reply:
We are not updating constructors in the overload/trampoline case bc the generated method did not follow the COnstructor format. I need to see if that changes with the change to not use the generator anymore.